### ISSUE TITLE
What am I doing wrong?

### ISSUE THREAD

**Issue by 9mm** [L3MON4D3/LuaSnip#978]
I think I'm very clearly doing something wrong, as snippets just dont work at all.

If i go to a new file, `:setf javascript` and I type: `xxx` and hit tab, the cursor doesnt move at all, and no snippet appears, as if it were a `<NOP>`. `<S-Tab>` does the same thing.

```
-- https://github.com/L3MON4D3/LuaSnip/blob/master/DOC.md#snipmate
-- examples: https://github.com/honza/vim-snippets

return {
  'L3MON4D3/LuaSnip',
  enabled = true,
  version = '2.*',
  build = 'make install_jsregexp',
  config = function()
    local ls = require('luasnip')

    ls.setup({})

    require('luasnip.loaders.from_snipmate').lazy_load({
      paths = '~/.config/nvim/snippets',
    })

    vim.keymap.set({'i'}, '<Tab>', function() ls.expand() end, { silent = true })
    vim.keymap.set({'i', 's'}, '<Tab>', function() ls.jump(1) end, { silent = true })
    vim.keymap.set({'i', 's'}, '<S-Tab>', function() ls.jump(-1) end, { silent = true })
    vim.keymap.set({'i', 's'}, '<C-e>', function()
      if ls.choice_active() then
        ls.change_choice(1)
      end
    end, { silent = true })
  end
}
```

and in my `~/.config/nvim/snippets/javascript.snippets`:

```
snippet xxx
	something
```

**Comment by L3MON4D3** [comment]
Can't spot anything wrong with it, you could try the following:
* set a high log-level before calling `lazy_load` (`ls.log.set_loglevel("info")`), and then check the log (`ls.log.open()`). It should contain messages about lazy-loading some files, and actually loading snippets from specific filetypes.
* Immediately open a file with filetype `javascript`
* Set filetype via `set filetype javascript`, I have never tried whether `setf` also calls the autocommands registered for `Filetype`
* Make sure the code is executed, by printing something

**Comment by 9mm** [comment]
`set filetype javascript` oddly doesn't work... `E518: Unknown option: javascript`.

When I use `set filetype=javascript` however, it logs this:

```
INFO  | snipmate-loader: Registering lazy-load-snippets:
      | {
      |   javascript = { "/Users/zesty/.config/nvim/snippets/javascript.snippets" }
      | }
INFO  | snipmate-loader: Loading lazy-load-snippets for filetype ``
INFO  | lua-loader: Loading lazy-load-snippets for filetype ``
INFO  | vscode-loader: Loading lazy-load-snippets for filetype ``
INFO  | lua-loader: Loading lazy-load-snippets for filetype `javascript`
INFO  | snipmate-loader: Loading lazy-load-snippets for filetype `javascript`
INFO  | snipmate-loader: Adding 1 snippets and 0 autosnippets for filetype `javascript` from /Users/zesty/.config/nvim/snippets/javascript.snippets
INFO  | vscode-loader: Loading lazy-load-snippets for filetype `javascript`
INFO  | lua-loader: Loading lazy-load-snippets for filetype `fidget`
INFO  | snipmate-loader: Loading lazy-load-snippets for filetype `fidget`
INFO  | vscode-loader: Loading lazy-load-snippets for filetype `fidget`
INFO  | lua-loader: Loading lazy-load-snippets for filetype `cmp_menu`
INFO  | snipmate-loader: Loading lazy-load-snippets for filetype `cmp_menu`
INFO  | vscode-loader: Loading lazy-load-snippets for filetype `cmp_menu`
```

And the behavior is the same, if I push `xxx` it basically <NOP>'s as if my tab key is broken

**Comment by 9mm** [comment]
I have also disabled every single plugin and the same thing occurs 🤔

**Comment by 9mm** [comment]
Interestingly, if I do `<C-k>` it works fine. It really doesn't work with `<Tab>` however... even if all other plugins are off.

**Comment by 9mm** [comment]
After looking at someone else's config I changed `jump` to `expand_or_jump` and now it seems to work (so far)... maybe that should be updated in docs if thats not how those functions are supposed to be used?

I will keep testing

**Comment by L3MON4D3** [comment]
Oh, I missed that `<tab>`-binding :sweat_smile:
Yeah, you probably want only `expand_or_jump` to be on `<tab>`, not `jump` alone as well.
What is your config now?

**Comment by 9mm** [comment]
Here's my current full config where the <Tab> key is completely <nop>/dead. The only thing it CAN do now is expand snippets 💀

```
-- https://github.com/L3MON4D3/LuaSnip/blob/master/DOC.md#snipmate
-- examples: https://github.com/honza/vim-snippets

return {
  'L3MON4D3/LuaSnip',
  enabled = false,
  event = 'VeryLazy',
  version = '2.*',
  build = 'make install_jsregexp',
  config = function()
    local ls = require('luasnip')

    ls.setup({})

    require('luasnip.loaders.from_snipmate').lazy_load({
      paths = '~/.config/nvim/snippets',
    })

    vim.keymap.set('i', '<Tab>', function() ls.expand_or_jump() end, { silent = true })
    vim.keymap.set('i', '<S-Tab>', function() ls.jump(-1) end, { silent = true })
    vim.keymap.set('i', '<C-e>', function()
      if ls.choice_active() then
        ls.change_choice(1)
      end
    end, { silent = true })
  end
}
```

**Comment by L3MON4D3** [comment]
Ahhh, it's probably that the `expand_or_jump`-mapping is only bound in INSERT, as opposed to INSERT and SELECT, which means that it won't work if the insertNode contains text which is selected.
tldr, just do `{"i", "s"}` instead of `"i"`, for all mappings actually.

**Comment by 9mm** [comment]
I don't think its that, let me explain again:

* I disabled all other plugins
* I ensured nothing is conflicting with `<Tab>`:

The only thing listed is `<Tab> * <Lua 336 ...luasnip.lua:19>`

```
-- https://github.com/L3MON4D3/LuaSnip/blob/master/DOC.md#snipmate
-- examples: https://github.com/honza/vim-snippets

return {
  'L3MON4D3/LuaSnip',
  enabled = true,
  event = 'VeryLazy',
  version = '2.*',
  build = 'make install_jsregexp',
  config = function()
    local ls = require('luasnip')

    ls.setup({})

    require('luasnip.loaders.from_snipmate').lazy_load({
      paths = '~/.config/nvim/snippets',
    })

    vim.keymap.set({'i', 's'}, '<Tab>', function() ls.expand_or_jump() end, { silent = true })
    vim.keymap.set({'i', 's'}, '<S-Tab>', function() ls.jump(-1) end, { silent = true })
    vim.keymap.set({'i', 's'}, '<C-e>', function()
      if ls.choice_active() then
        ls.change_choice(1)
      end
    end, { silent = true })
  end
}
```

When I have `<Tab>` set to `expand_or_jump()`, and the config above, I cannot use the tab key at all, whether its insert mode, select mode, etc. It behaves as if its a <NOP> or my key is actually broken.

The *only* thing that the tab key works for, is expanding a snippet.

So if I have a snippet for `xxx` -> `something`

I can do this:

`xxx|` = tab WILL expand
`x|` = tab will not move, or insert tabs. its broken. Likewise, I can just have a completely empty file:

`|` and tab will still just not insert tabs

-----

The other option is this:

```
    vim.keymap.set({'i', 's'}, '<Tab>', function() ls.expand() end, { silent = true })
    vim.keymap.set({'i', 's'}, '<Tab>', function() ls.jump(1) end, { silent = true })
    vim.keymap.set({'i', 's'}, '<S-Tab>', function() ls.jump(-1) end, { silent = true })
    vim.keymap.set({'i', 's'}, '<C-e>', function()
```

Notice the extra keymap and using `expand()` instead of `expand_or_jump()`

Using this way, it's identical to the first problem, except now in addition it wont tab EVEN if there is an `xxx` snippet. In other words, the tab is now useless in insert mode, whether theres a snippet or not.

**Comment by L3MON4D3** [comment]
Ah, you want some kind of supertab-mapping, okay.
Luasnip doesn't have something like this built in, but you can either find something like this in other ppls config, or cobble something together yourself: `ls.expand_or_jumpable()` (I think that's the name, check doc if it doesn't work) returns whether you can jump or expand right now, and you could insert a tab, or expand_or_jump based on that.
I think you should find lots of examples by searching for "if ls.expand_or_jumpable() then" on github

**Comment by 9mm** [comment]
awesome thanks!

**Comment by 9mm** [comment]
Ill close this and paste my answer here when i get it

**Comment by 9mm** [comment]
Ok this is how I solved it:

```lua
-- examples: https://github.com/honza/vim-snippets/tree/master/snippets
-- snippet format: https://github.com/garbas/vim-snipmate/blob/master/doc/SnipMate.txt#L228

return {
  'L3MON4D3/LuaSnip',
  enabled = true,
  event = 'VeryLazy',
  version = '2.*',
  build = 'make install_jsregexp',
  config = function()
    local ls = require('luasnip')

    -- https://github.com/L3MON4D3/LuaSnip/blob/master/DOC.md#config-options
    ls.setup({
      update_events = 'TextChanged,TextChangedI',
    })

    require('luasnip.loaders.from_snipmate').lazy_load({
      paths = '~/.config/nvim/snippets',
    })

    local function interp(k) vim.api.nvim_feedkeys(vim.api.nvim_replace_termcodes(k, true, false, true), 'n', false) end

    vim.keymap.set({'i', 's'}, '<Tab>', function()
      if ls.expand_or_jumpable() then ls.expand_or_jump()
      else interp('<Tab>') end
    end)

    vim.keymap.set({'i', 's'}, '<S-Tab>', function()
      if ls.jumpable(-1) then ls.jump(-1)
      else interp('<S-Tab>') end
    end)
  end
}
```

**Comment by 9mm** [comment]
Now the problem is getting it to work with tabout, but I'm doing that here (in case anyone else needs this)

https://github.com/abecodes/tabout.nvim/issues/57

**Comment by L3MON4D3** [comment]
Looks good 👍👍

**Comment by 9mm** [comment]
@L3MON4D3 ok i realize this isn't your plugin (tabout) but at a higher level... I'm confused about the best way to set this up. So far after trying about an hour I am not getting Tabout+Luasnip to interact.

My main question is...

1. should tabout config go in luasnip maps, or luasnip go in tabout config?

It seems the latter is easier so thats what I started with. Inside Luasnip I just didn't define any mappings at all, and in tabout I did this:

```lua
    local ls = require('luasnip')

    local function replace_keycodes(c)
      return vim.api.nvim_replace_termcodes(c, true, true, true)
    end

    function _G.tab_binding()
      if ls.expand_or_jumpable() then
        return replace_keycodes('<Plug>(luasnip-expand-or-jump)')
      else
        return replace_keycodes('<Plug>(Tabout)')
      end
    end

    function _G.s_tab_binding()
      if ls.jumpable(-1) then
        --
      else
        return replace_keycodes('<Plug>(TaboutBack)')
      end
    end

    vim.api.nvim_set_keymap('i', '<Tab>', 'v:lua.tab_binding()', { expr = true })
    vim.api.nvim_set_keymap('i', '<S-Tab>', 'v:lua.s_tab_binding()', { expr = true })
```

But when I do this, the snippet expansion actually just prints out `<Plug>(luasnip-expand-or-jump)` (lol)

I've also tried every combination of calling the function directly, etc, that I can think of. It has any number of outcomes...

1. it inserts `v:null` for the snippet
2. it has a stack trace exception

Is there a way to do this cleanly/properly which isn't a bunch of antipatterns? Something about the way I'm doing this just feels like it's the completely wrong way, even if i get it working somehow.

I'm assuming Im supposed to leverage the <Plug> stuff...?
